File: frontend/src/utils/cache.ts
/**
 * 简单的内存缓存工具
 */
interface CacheItem<T> {
  data: T
  timestamp: number
  expiresIn: number
}

class SimpleCache {
  private cache = new Map<string, CacheItem<any>>()

  /**
   * 设置缓存
   * @param key 缓存键
   * @param data 缓存数据
   * @param expiresIn 过期时间（毫秒），默认 5 分钟
   */
  set<T>(key: string, data: T, expiresIn: number = 5 * 60 * 1000): void {
    this.cache.set(key, {
      data,
      timestamp: Date.now(),
      expiresIn,
    })
  }

  /**
   * 获取缓存
   * @param key 缓存键
   * @returns 缓存数据，如果过期或不存在则返回 null
   */
  get<T>(key: string): T | null {
    const item = this.cache.get(key)
    if (!item) {
      return null
    }

    const now = Date.now()
    if (now - item.timestamp > item.expiresIn) {
      // 过期，删除
      this.cache.delete(key)
      return null
    }

    return item.data as T
  }

  /**
   * 删除缓存
   */
  delete(key: string): void {
    this.cache.delete(key)
  }

  /**
   * 清空所有缓存
   */
  clear(): void {
    this.cache.clear()
  }

  /**
   * 检查缓存是否存在且未过期
   */
  has(key: string): boolean {
    const item = this.cache.get(key)
    if (!item) {
      return false
    }

    const now = Date.now()
    if (now - item.timestamp > item.expiresIn) {
      this.cache.delete(key)
      return false
    }

    return true
  }

  /**
   * 清除所有以指定前缀开头的缓存
   */
  clearByPrefix(prefix: string): void {
    const keysToDelete: string[] = []
    // 遍历所有缓存键
    for (const key of this.cache.keys()) {
      if (key.startsWith(prefix)) {
        keysToDelete.push(key)
      }
    }
    // 删除匹配的缓存
    keysToDelete.forEach(key => this.cache.delete(key))
  }
}

// 导出单例
export const cache = new SimpleCache()

File: backend/app/utils/cache.py
"""
简单的内存缓存工具
用于缓存API响应，减少数据库查询
"""
from typing import Optional, Dict, Any
from datetime import datetime, timedelta
import threading
import logging

logger = logging.getLogger(__name__)


class SimpleCache:
    """简单的内存缓存实现"""
    
    def __init__(self):
        self._cache: Dict[str, Dict[str, Any]] = {}
        self._lock = threading.Lock()
    
    def get(self, key: str) -> Optional[Any]:
        """获取缓存值"""
        with self._lock:
            if key not in self._cache:
                return None
            
            item = self._cache[key]
            # 检查是否过期
            if datetime.now() > item['expires_at']:
                del self._cache[key]
                return None
            
            return item['value']
    
    def set(self, key: str, value: Any, expires_in: int = 300) -> None:
        """
        设置缓存值
        
        Args:
            key: 缓存键
            value: 缓存值
            expires_in: 过期时间（秒），默认5分钟
        """
        with self._lock:
            expires_at = datetime.now() + timedelta(seconds=expires_in)
            self._cache[key] = {
                'value': value,
                'expires_at': expires_at
            }
    
    def delete(self, key: str) -> None:
        """删除缓存"""
        with self._lock:
            if key in self._cache:
                del self._cache[key]
    
    def clear(self, prefix: Optional[str] = None) -> None:
        """
        清除缓存
        
        Args:
            prefix: 如果提供，只清除以该前缀开头的缓存
        """
        with self._lock:
            if prefix:
                keys_to_delete = [k for k in self._cache.keys() if k.startswith(prefix)]
                for key in keys_to_delete:
                    del self._cache[key]
            else:
                self._cache.clear()
    
    def clear_expired(self) -> None:
        """清除所有过期的缓存"""
        with self._lock:
            now = datetime.now()
            keys_to_delete = [
                key for key, item in self._cache.items()
                if now > item['expires_at']
            ]
            for key in keys_to_delete:
                del self._cache[key]


# 全局缓存实例
cache = SimpleCache()

